File: src/sorting/cpp/selection sort/src/selection_sort.cpp
#include "selection_sort.h"
#include <algorithm>
#include <vector>

template <class T> void selection_sort(std::vector<T> &arr) {
  for (auto it = arr.begin(); it != arr.end(); ++it) {
    auto const selection = std::min_element(it, arr.end());
    std::iter_swap(selection, it);
  }
}

template void selection_sort<int>(std::vector<int> &arr);
template void selection_sort<float>(std::vector<float> &arr);
template void selection_sort<double>(std::vector<double> &arr);
template void selection_sort<char>(std::vector<char> &arr);

File: src/sorting/python/selection_sort/src/selection_sort.py
"""Selection sort algorithm implementation."""

from typing import List, TypeVar

T = TypeVar("T")


def selection_sort(array: List[T]) -> List[T]:
    """
    Sort an array using the selection sort algorithm.

    This algorithm divides the input into a sorted and unsorted region,
    repeatedly selecting the smallest element from the unsorted region
    and moving it to the end of the sorted region.

    Time complexity: O(n²)
    Space complexity: O(1)

    Args:
        array: The list of comparable elements to sort.

    Returns:
        The same list, sorted in ascending order.
    """
    for i in range(len(array)):
        min_idx = i

        for j in range(i + 1, len(array)):
            if array[j] < array[min_idx]:
                min_idx = j

        array[i], array[min_idx] = array[min_idx], array[i]

    return array
